gapis/resolve/dependencygraph2/dependency_graph.go

```diff
@@ -0,0 +1,425 @@
+	Capture() *capture.Capture
+
+	// GetCommand returns the command identified by the given CmdID
+	GetCommand(api.CmdID) api.Cmd
+
+	// InitialCommands returns the initial commands, which reconstruct the
+	// initial state before the first command in the capture.
+	InitialCommands() []api.Cmd
+
+	// InitialCmdRanges returns the memory ranges used by the initial commands.
+	InitialCmdRanges() interval.U64RangeList
+
+	// Config returns the config used to create this graph
+	Config() DependencyGraphConfig
+}
+
+type obsNodeIDs struct {
+	numReads int
+	nodeIDs  []NodeID
+}
+
+type dependencyGraph struct {
+	capture           *capture.Capture
+	cmdNodeIDs        *api.SubCmdIdxTrie
+	initMemoryNodeIDs obsNodeIDs
+	initCmdObsNodeIDs []obsNodeIDs
+	cmdObsNodeIDs     []obsNodeIDs
+	initialCommands   []api.Cmd
+	initialCmdRanges  interval.U64RangeList
+	nodes             []Node
+	numDependencies   int
+	dependenciesFrom  [][]NodeID
+	dependenciesTo    [][]NodeID
+
+	config DependencyGraphConfig
+}
+
+const initCmdBit = api.CmdID(1 << 62)
+
+// newDependencyGraph constructs a new dependency graph
+func newDependencyGraph(ctx context.Context, config DependencyGraphConfig,
+	c *capture.Capture, initialCmds []api.Cmd, initialCmdRanges interval.U64RangeList) *dependencyGraph {
+	g := &dependencyGraph{
+		capture:          c,
+		cmdNodeIDs:       new(api.SubCmdIdxTrie),
+		initialCommands:  initialCmds,
+		initialCmdRanges: initialCmdRanges,
+		config:           config,
+	}
+	numCmds := len(initialCmds) + len(c.Commands)
+	numObservations := countObservations(initialCmds) + countObservations(c.Commands)
+	numNodes := numCmds + numObservations
+	g.nodes = make([]Node, 0, numNodes)
+	g.initCmdObsNodeIDs = make([]obsNodeIDs, len(initialCmds))
+	for i, cmd := range initialCmds {
+		g.initCmdObsNodeIDs[i] = g.addCmdNodes(api.CmdID(i)|initCmdBit, cmd.Extras().Observations())
+	}
+	g.cmdObsNodeIDs = make([]obsNodeIDs, len(c.Commands))
+	for i, cmd := range c.Commands {
+		g.cmdObsNodeIDs[i] = g.addCmdNodes(api.CmdID(i), cmd.Extras().Observations())
+	}
+	g.dependenciesFrom = make([][]NodeID, numNodes)
+	return g
+}
+
+// NumNodes returns the number of nodes in the graph
+func (g *dependencyGraph) NumNodes() int {
+	return len(g.nodes)
+}
+
+// NumDependencies returns the number of dependencies (edges) in the graph
+func (g *dependencyGraph) NumDependencies() int {
+	return g.numDependencies
+}
+
+// GetNode returns the node data associated with the given NodeID
+func (g *dependencyGraph) GetNode(nodeID NodeID) Node {
+	if nodeID >= NodeID(len(g.nodes)) {
+		return nil
+	}
+	return g.nodes[nodeID]
+}
+
+// GetNodeID returns the NodeID associated with given node data
+func (g *dependencyGraph) GetNodeID(node Node) NodeID {
+	if cmdNode, ok := node.(CmdNode); ok {
+		index := cmdNode.Index
+		if len(index) == 0 {
+			return NodeNoID
+		}
+		if !g.config.IncludeInitialCommands && !api.CmdID(index[0]).IsReal() {
+			return NodeNoID
+		}
+		if g.config.MergeSubCmdNodes {
```

Comment:
Do we actually support subcommands? It looks like we never put a command id with len() > 1 to the `cmdNodeIDs` trie.

Reply:
At this point, the dependency graph and DCE do not handle subcommands. I hope to add this capability soon, but I am worried about memory and performance. This is likely to at least double the number of nodes--each `vkCmd...` is will have a node when recorded to the command buffer, and a node each time the command buffer is submitted.